### ISSUE TITLE
Make graphs / pages shareable via URL

### ISSUE THREAD

**Issue by ahpook** [oss-aspen/8Knot#706]
**Describe the bug**
I can't find a way to share a specific page of graphs with other people - Ideally there'd be a way to copy/paste the url to a specific graph that preserves my current view options, but it'd also be helpful to share the default results for a specific repo/dataset.

**To Reproduce**
1. Load up the data for a specific repository
2. Copy the URL
3. Paste the URL into a new anonymous browser window

**Expected behavior**
The new browser window should have the same view as the original one

**Comment by cdolfi** [comment]
@ahpook Great idea! We will keep that in mind. Have you checked out using the "download graphs as png" functionality on the graphs for the time being?

**Comment by ahpook** [comment]
@cdolfi Yes, the static download is OK - but ideally I'd like to send someone a link to a filtered set of repositories, so they can start exploring the same info.

**Comment by cdolfi** [comment]
Yes 100% agree. We will loop back around on this one the redesign(https://www.figma.com/file/e9ylKHtU7YgWyuXBkI9bm0/Augur-8Knot?type=design&node-id=1%3A63&mode=design&t=tZrTpxVfy5oBKpFO-1) is finished. If it ends up being something a GH engineer would like to work on in the mean time would be happy to talk to them!

**Comment by MoralCode** [comment]
In addition to the child issue that is potentially a blocker for this, I think we also need to be very careful with how we design the url.

Our only place to store information when a link is shared is the url. Fundamentally for this to work, we need two things to be available:
1. a way to identify the specific graph that is being shared. This is already accomplished and shipping in the code using the path (graph category) and fragment (specific graph) that were added as part of the redesign.
2. a way to identify the exact set of repos and orgs that were part of the search that generated the graph. This is the major piece we do not currently have.


In addition to providing the information necessary to take the user to the correct graph, *how* we provide it also matters.

If we simply list each repo as an org/repo string:
- we may not be able to differentiate github repos vs other (i.e. gitlab/forgejo) repos
- for searches with large numbers of repos/orgs in them, do we want to cap the number so that we dont run into issues with maximum URL lengths?

If we list the repos by ID, then the url wont be portable across instances. This was never something that was asked for, but the point is that we would want to document this tradeoff.

We *could* introduce a new database table or other form of persistent storage in 8knot that allows us to create a "share ID". This would be a short number that gets internally mapped to the set of repos that were part of that search. That would also need to be per-instance, but it would eliminate the need to encode all the data in the URL and may provide more flexibility for future migrations (i.e. if needed we could ship changes to 8knot that map the share ID to other things that are functionally equivalent to what was shared).

Base64 or otherwise encoding the list may also give us slightly more room to add values if we need space.

I think this needs a more thorough design discussion to make sure we get this right so that we can release a system where we can confidently commit to not breaking links that people share.


Another general thing to think about is how we want to handle cases where a visualization or category that is being shared no longer exists in 8knot (i.e. the heatmaps or affiliations page getting disabled, us renaming the category names or changing the graph IDs)

**Comment by cdolfi** [comment]
@EngCaioFonseca Can you describe the high level implementation here?

**Comment by EngCaioFonseca** [comment]
<img width="1398" height="1206" alt="Image" src="https://github.com/user-attachments/assets/1b18b289-140b-4930-b18c-b3afe1652faa" />

**Comment by MoralCode** [comment]
Criteria i think that the newly-added parameters in the URL need to be able to meet:

- Able to identify the exact set of repos and orgs that were part of the search that generated the graph
- able to identify the exact graph that was being shared
- Able to scale to accommodate the future addition of repositories from other forges (as the upstream collection tooling supports more and more forge sources)
- Able to handle potentially changing groupings of repos (maybe a repo moves from one org to another, or the search is based on a user-defined group)
- shared URLs also preserve settings (i.e. bots disabled) that were used in the original search (user looking at the link should see the exact graph that was shared (plus any new data since that graph was created)
- The format should be flexible and allow us to change it in the future - just in case

Important implementation considerations:
- Allow/document a path for instance admins to not break their users links when migrating or updating their instance (either plan for instances to migrate their links/IDs to a new instance, or support links that work across instances)
- Plan for how to handle common/foreseeable edge cases (what if one repo/group in the search set is no longer present on the instance? how should this be communicated to users?)
- in addition to versioning or some other kind of support for changing the url format, there should be an implementation for such potential migration path, such as giving 8Knot the ability to redirect older versions of the link format to newer ones, or architect it such that there is a spot where we COULD increment a version and have different logic if we wanted to. i.e. where in the code WOULD we do a migration if we wanted to?


## Misc/nice to haves

This list also doesnt account for potentially wanting to show the specific timeframe within a graph that was shared. If we want this feature, there are additional considerations:
- If I share a link today showing the "Last 30 Days," and it is opened again in three months, should the person who opens it see the data from when I sent it or the new last 30 days?


If we want nice previews (predominantly useful for chat apps) we would likely want to support dynamic opengraph tags - essentially this is the HTML page metadata that allows slack to unfurl a link preview when it is shared, rather than just a long URL with no context. This could help achieve user intuitive design without having to make the URL itself human readable. 

Shareable URLs may also permanently prevent 8Knot from supporting analysis of private repos

**Comment by MoralCode** [comment]
Part of me is interested in potentially making this url sharing its own little subsystem.

i.e. we have a "share" button on each graph. Clicking that will trigger some code that generates a link (i like the format `https://eightknot.osci.io/s/Ab12Gz`) with a database-backed sharing ID. the sub-path of `/s/` would make the routing super easy, and the user-triggered link generation could create a new entry in 8Knot's database (we may need to create a new schema for this) to store the searched repos and graph that were linked to from that newly generated ID.

We could also find a way to encode the data in the url though

**Comment by pdurbin** [comment]
This all sounds very fancy and nice, but what about the simple use case of a single repo? Would that be a good place to start? I mentioned the single repo use case over in this issue, which got closed as a duplicate:

- #739

**Comment by cdolfi** [comment]
@pdurbin We want to make sure if we release a feature, it works with the functionality of the current application. As well as not having issues with breaking links in the future if a single repo approach does not scale to repos and orgs. If you would like us to discuss the implementation of this on a separate issue happy to make a new one

**Comment by pdurbin** [comment]
> If you would like us to discuss the implementation of this on a separate issue happy to make a new one

@cdolfi Sure, please go ahead and I'll jump in! Thanks! ❤️

**Comment by MoralCode** [comment]
hmm, the URL format in the screenshot in that issue looks interesting. 8Knot currently already uses the url path for the category/page of graphs you are looking at so it wont work in the main URL namespace, but i wonder if we could have two different styles of shareable URL (building on my conversation with Caio today):

- eightknot.osci.io/s/W3t8gX (sharing a multiple repos/orgs, where the trailing characters either map to an internal database of the things you searched for, or somehow directly encodes it)
- eightknot.osci.io/r/github/org/repo (sharing stats for a single repo/org)

the single repo option could also unlock some interesting usecases, like integrating with shields.io to maybe show off a particular metric in your org's README or something
